barterdude/hooks/metrics/prometheus/__init__.py

```diff
@@ -0,0 +1,106 @@
+import time
+import copy
+from aiohttp import web
+from typing import Optional
+
+from barterdude import BarterDude
+from barterdude.hooks import HttpHook
+from barterdude.hooks.metrics.prometheus.definition import Definition
+
+try:
+    from prometheus_client import (
+        CollectorRegistry,
+        generate_latest,
+        CONTENT_TYPE_LATEST
+    )
+except ImportError:
+    print("""
+    Please install extra dependency with:
+        `pip install barterdude[prometheus]`
+    """)
+
+
+class Prometheus(HttpHook):
+
+    MESSAGE_UNITS = "messages"
+    NAMESPACE = "barterdude"
+    D_BEFORE_CONSUME = "before_consume"
+    D_SUCCESS = "success"
+    D_FAIL = "fail"
+    D_TIME_MEASURE = "time_measure"
+
+    def __init__(
+        self,
+        barterdude: BarterDude,
+        labels: dict,
+        path: str = "/metrics",
+        registry: CollectorRegistry = CollectorRegistry(),
+        definition: Definition = Definition()
+    ):
+        self.__registry = registry
+        self.__labels = labels
+        self.__labelnames = list(labels.keys())
+        self.__labelvalues = list(labels.values())
+        self._d = definition
+        self._time_bucket = {}
+        self.__prepare_metrics()
+        super(Prometheus, self).__init__(barterdude, path)
+
+    def __prepare_metrics(self):
+        default_kwargs = dict(
+            labelnames=self.__labelnames,
+            namespace=self.NAMESPACE,
+            unit=self.MESSAGE_UNITS,
+            registry=self.__registry
+        )
+        self._d.prepare_before_consume(
+            self.D_BEFORE_CONSUME, copy.deepcopy(default_kwargs)
+        )
+        self._d.prepare_on_complete(
+            self.D_SUCCESS, copy.deepcopy(default_kwargs)
+        )
+        self._d.prepare_on_complete(self.D_FAIL, copy.deepcopy(default_kwargs))
+        self._d.prepare_time_measure(
+            self.D_TIME_MEASURE, copy.deepcopy(default_kwargs)
+        )
+
+    async def __compute_hash(self, message: dict):
+        return id(message)
+
+    async def before_consume(self, message: dict):
+        hash_message = await self.__compute_hash(message)
+        self._d.senders[self.D_BEFORE_CONSUME].labels(
+            **self.__labels
+        ).inc()
+        self._time_bucket[hash_message] = time.time()
+
+    async def _on_complete(self,
+                           message: dict,
+                           state: str,
+                           error: Optional[Exception] = None):
+
+        final_time = time.time()
+        hash_message = await self.__compute_hash(message)
```

Comment:
I would inline these variables.

Reply:
final_time is necessary because I want the processing time only of the wrapped function... hash_message I prefer to explicit this name because id(message) is not so easy to understand